### PR TITLE
Add SecureSocket change to the module

### PR DESCRIPTION
**Pull request by lahirusamith**
## Learning
 can not use
 `Type CertKey Record {|
    string certFile;
    string keyFile;
    string keyPassword?;
}`
to add combination of certificate and private key of the client for the proposed "key" field in SecureSocket because, according to the [mysql connector/j documentation ](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-using-ssl.html), you can only provide the keystore to the connector.

Also,

can not use
`string` 
only field for the proposed "cert" field in SecureSocket, as per the same documentation mentioned above.

### CODE REVIEW

**Review by daneshk — APPROVED** [review]

File: mysql-ballerina/client.bal [hunk 1]
```diff
@@ -169,13 +169,12 @@ public type SSLMode SSL_PREFERRED|SSL_REQUIRED|SSL_VERIFY_CERT|SSL_VERIFY_IDENTI
...
-# + trustCertKeystore - Keystore configuration of the trust certificates
-#
-public type SSLConfig record {|
+# + key - Keystore configuration of the client certificates
+# + cert - Keystore configuration of the trust certificates
+public type SecureSocket record {|
     SSLMode mode = SSL_PREFERRED;
-    crypto:KeyStore clientCertKeystore?;
-    crypto:KeyStore trustCertKeystore?;
+    crypto:KeyStore key?;
```
Comment (ldclakmal) [inline mysql-ballerina/client.bal:176]: Can't we make this mandatory?
```suggestion
    crypto:KeyStore key;
```

File: mysql-ballerina/client.bal [hunk 2]
```diff
@@ -151,7 +151,7 @@ type ClientConfiguration record {|
 # + socketTimeout - Socket timeout (in seconds) during the read/write operations with mysql server,
 #                            0 means no socket timeout
 public type Options record {|
-    SSLConfig? ssl = {};
+    SecureSocket? ssl = {};
```
Comment (ldclakmal) [inline mysql-ballerina/client.bal:154]: Can't we make this optional?
```suggestion
    SecureSocket ssl?;
```

File: mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java [hunk 3]
```diff
@@ -59,34 +59,38 @@ public static long getTimeout(Object secondsDecimal) {
...
-        if (sslConfig == null) {
+    public static void addSSLOptions(BMap secureSocket, BMap<BString, Object> options) {
+        if (secureSocket == null) {
             options.put(Constants.DatabaseProps.SSL_MODE, Constants.DatabaseProps.SSL_MODE_DISABLED);
         } else {
-            BString mode = sslConfig.getStringValue(Constants.SSLConfig.MODE);
-            if (mode.getValue().equalsIgnoreCase(Constants.SSLConfig.VERIFY_CERT_MODE)) {
+            BString mode = secureSocket.getStringValue(Constants.SecureSocket.MODE);
+            if (mode.getValue().equalsIgnoreCase(Constants.SecureSocket.VERIFY_CERT_MODE)) {
                 mode = Constants.DatabaseProps.SSL_MODE_VERIFY_CA;
```
Comment (ldclakmal) [inline mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java:68]: AFAIU, we simply set the passed SSL mode from Ballerina side into Java options. Therefore, can't we update the Ballerina constant `public const SSL_VERIFY_CERT = "VERIFY_CERT";` as `public const SSL_VERIFY_CA = "VERIFY_CA";` and remove this if condition?
@lahirusamith @daneshk

Comment (lahirusamith) [reply mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java:68]: @ldclakmal, Yes It seems we can. As per the [documentation ](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-connp-props-security.html). @daneshk WDYT?

Comment (daneshk) [reply mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java:68]: +1

File: mysql-ballerina/client.bal (hunk 1 above)
Comment (lahirusamith) [reply mysql-ballerina/client.bal:176]: we can not make that field mandatory. according to the mysql documentation, there are several ssl mode available and key and cert is depended on that mode selection.

Comment (ldclakmal) [reply mysql-ballerina/client.bal:176]: @lahirusamith Just out of curiosity, what is the mode (use case), where we have to use `cert` field without `key` field?

File: mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java (hunk 3 above)
Comment (lahirusamith) [reply mysql-native/src/main/java/org/ballerinalang/mysql/Utils.java]: changed in https://github.com/ballerina-platform/module-ballerinax-mysql/pull/96/commits/45b458de1d5df35268be9b467362e48c0d0fcc17

File: mysql-ballerina/client.bal (hunk 2 above)
Comment (lahirusamith) [reply mysql-ballerina/client.bal]: made Optional in https://github.com/ballerina-platform/module-ballerinax-mysql/pull/96/commits/45b458de1d5df35268be9b467362e48c0d0fcc17

File: mysql-ballerina/client.bal (hunk 1 above)
Comment (lahirusamith) [reply mysql-ballerina/client.bal:176]: both key and cert may be optional for SSL_PREFERRED.

**Pull request merged**